keps/sig-cluster-lifecycle/20190227-artifact-generation.md

```diff
@@ -0,0 +1,197 @@
+- Artifact generation is independent of source code compilation. 
+
+### Non-Goals
+
+- A change in the release process. see this [kep](https://github.com/kubernetes/enhancements/pull/843)
+    - For example, package signing and distribution are not covered in this document and are covered in the above KEP.
+
+- To fundamentally change the packages and their dependencies. All packages should remain consistent with what we build today. 
+
+## Proposal
+
+As outlined in the following issue:
+Base ticket: https://github.com/kubernetes/kubernetes/issues/71677
+Move all packaging into kubernetes/kubernetes repo and create the following 3 building blocks to support it.
+Create a configuration yaml file to enable the configuring of the built packages.
+Create a Go lang app to parse yaml config and submit simple commands to the container.
+Build inside a docker container based off FPM.
+
+### User Stories
+
+#### Story 1
+
+As a developer I would like to be able to install the entire set of alpha or beta (.deb/.rpm) packages to stand up a development cluster.  These packages should be equivalent to the future release artifacts.  
+
+#### Story 2
+
+As a third party vendor, I will be able to package kubernetes using the same tooling that kubernetes’ uses. 
+
+#### Story 3
+
+As a release manager I would like to enable simple and easy package generation that the community can manage. 
+
+### Implementation Details/Notes/Constraints [optional]
+
+- Use an fpm based container to build packages
+    - Nfpm was originally investigated but lacked features
+- Use yaml to configure
+- Create a go binary that can consume the yaml, build the right file structure and issue commands to the image to build the image.
+
+### Risks and Mitigations
+
+Risk: Confusion as to what to use when. Mitigation: Coordination across groups to ensure that the changes made in k/k are consumed by the release team. 
+## Design Details
+
+### Test Plan
+
+Ideally these new packages will get tested inside the test pipeline. When KIND goes to run kubeadm/kubelet/kubectl/etc tests, it should install the packages generated by these tools as they will be contained within the kubernetes/kubernetes repository.
+
+This will give us clear paths to the package generated and the package consumed when testing enabling a more confident release.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+
+### Graduation Criteria
+
+N/A
+
+#### Examples
+
+1:
+An independent developer wanting to generate packages in their own fork would be able to edit the config yaml in their fork, override a dependency name and then issue a command to the go packaging binary to generate the packages.
+`[go packing binary] --config=myconfig.yaml`
+
+2:
+Bazel will be able to update the configuration yaml to dynamically update the configuration before simply executing commands directly against the go packing binary.
+
```

Comment:
should this section be N/A and the contents be move as a User Story?

Reply:
I'm not really sure. I felt like the examples where a bit of a rehash of user stories as well as implementation details was a rehash of proposal. Sorry, this is my First KEP, I'm up for whatever you all think is best. I can do that. These are worded slightly differently...